File: project6/submit/C/task1.c
/**
 * CS333 Project 6
 * task1.c
 * Word counter that can count the number of occurrences of every word in a text file
 * 
 * Di Luo
 * 05/10/2020
 */

#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include "linkedlist.h"

typedef struct wordcounter {
    int wordcount;
    char word[50];
}WordCounter; 

// create a new WordCounter
WordCounter *wc_create(){
    WordCounter *wc = (WordCounter*)malloc(sizeof(WordCounter));
    wc->wordcount = 0;
    return wc;
}

// Process word in file
void process(char *word){
    int len = strlen(word);
    if (((int)word[len-1] < 97) || (int)word[len-1] > 122){
        word[len-1] = '\0';
    }
}

// Increment word count
void increment(WordCounter wc[], char word[], int *n){
    strcpy(wc[*n].word, word);
    wc[*n].wordcount = 1;
    (*n)++;
}

// Analyze the file with addWord 
void analyze(FILE *file, WordCounter wc[], int *n){
    char word[50];
    int found = 0;
    int test = 0;
    while (fscanf(file,"%s", word) == 1){
        int i = 0;
        while(word[i]){
            word[i] = tolower(word[i]);
            i++;
        }
        process(word);
        for (int j = 0; j < *n; j++){
            if (strcmp(wc[j].word, word) == 0){
                found = 1;
                wc[j].wordcount++;
            }
        }
        if (found == 0){
            increment(wc,word,n);
        }
        found = 0;
    }
}

// Comparator function in descending order
int comparator(const void *p, const void *q) {
    WordCounter a = *(WordCounter *)p;
    WordCounter b = *(WordCounter *)q;
    return (b.wordcount - a.wordcount);
}

// Sort the wordcount array by using qsort
void top20(WordCounter wc[], int *n){
    qsort((void *) wc, *n, sizeof(WordCounter), comparator);
    for (int i=0; i<20; i++){
        printf("%s:%d \n", wc[i].word, wc[i].wordcount);
    }
}

int main(int argc, char *argv[]){
    FILE *fp;
    WordCounter wc[100];
    int numElem = 0;
    if (argc == 1){
        printf("Missing input. Filename is needed.");
        exit(-1);
    }
    else{
        if ((fp = fopen(*++argv, "r"))== NULL){
            printf("The file %s cannot be opened\n", *argv);
            return 1;
        }
        else{
        analyze(fp,wc,&numElem);
        top20(wc,&numElem);
        fclose(fp);
        }
    }
}

File: project6/submit/JavaScript/task1.js
/**
 * CS333 Project 6
 * task1.js
 * Word counter that can count the number of occurrences of every word in a text file
 * 
 * Di Luo
 * 05/10/2020
 */

var map = new Map();

function analyze(filename){
    const fs = require('fs');
    var contents = fs.readFileSync(filename, 'utf8');
    contents = contents.toString();
    var list = contents.split(" ");
    var found = 0;
    for (var i = 0; i < list.length; i++){
        list[i] = list[i].replace(/\b[-.,()&$#!\[\]{}"']+\B|\B[-.,()&$#!\[\]{}"']+\b/g, "");
        list[i] = list[i].toLowerCase();
        if (map.has(list[i]) == true){
            map.set(list[i],map.get(list[i])+1);
            found = 1;
        }
        if (found == 0){
            map.set(list[i],1);
        }
        found = 0;
    }
}

function main(){
    analyze('wctest.txt');
    map[Symbol.iterator] = function*(){
        yield* [...this.entries()].sort((a,b)=> b[1]-a[1]);
    }
    let keys = [];
    for (let key of map){
        keys.push(key);
    }
    for (var i=0; i<20; i++){
        let x = keys[i][0]
        let y = keys[i][1]
        console.log(x + ' ' + y);
    }
}

main()